### PR TITLE
initial structure for MongoDB Atlas recevier

### PR DESCRIPTION
**Pull request by zenmoto**
**Description:**
This is the initial scaffolding without actual function for a new receiver to pull metrics and logs from [MongoDB Atlas](https://www.mongodb.com/cloud/atlas)

This PR only creates an empty receiver and stubs, actually fetching metrics will come next, followed by pulling logs.

### CODE REVIEW

File: cmd/configschema/go.mod [hunk 1]
```diff
@@ -626,6 +630,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/tcplo
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/udplogreceiver => ../../receiver/udplogreceiver
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver => ../../receiver/mongodbatlasreceiver
```
Comment (jpkrohling) [inline cmd/configschema/go.mod:633]: I feel bad for being extremely picky, but can you place this after `memcachedreceiver`?

File: receiver/mongodbatlasreceiver/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# MongoDB Atlas Receiver
+
+Receives metrics from [MongoDB Atlas](https://www.mongodb.com/cloud/atlas) 
+via their [monitoring APIs](https://docs.atlas.mongodb.com/reference/api/monitoring-and-logs/)
+
+Supported pipeline types: metrics
+
+## Getting Started
+
+- `public_key`
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/README.md:10]: Is there a recommendation about how those values should be stored?

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+func (s *MongoDBAtlasClient) Organizations(ctx context.Context) []*mongodbatlas.Organization {
+	allOrgs := make([]*mongodbatlas.Organization, 0)
+	page := 1
+
+	for {
+		orgs, hasNext, err := s.getOrganizationsPage(ctx, page)
+		page++
+		if err != nil {
+			// Error was reported in getOrganizationsPage
+			break // Stop, returning what we have (probably empty slice)
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:79]: Given that you are swallowing the error here, move the logging statement you have in the `getOrganizationsPage` to this place here. Make sure to use the `debug` level instead. Ideally, this failure would also be recorded as a gauge.

File: receiver/mongodbatlasreceiver/receiver.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		s.runner.Stop()
+	}
+	return nil
+}
+
+func (s *receiver) poll(start string, end string, resolution string) { //nolint
+	ctx := s.ctx
+	for _, org := range s.client.Organizations(ctx) {
+		// Metrics collection starts here
+		s.log.Info("fetch resources for MongoDB Organization", zap.String("org", org.Name))
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/receiver.go:91]: Debug instead?

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+func NewMongoDBAtlasClient(
+	publicKey string,
+	privateKey string,
+	log *zap.Logger,
+) (*MongoDBAtlasClient, error) {
+	t := digest.NewTransport(publicKey, privateKey)
+	tc, err := t.Client()
+	if err != nil {
+		return nil, err
```
Comment (rmfitzpatrick) [inline receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:40]: Wrapping with some context can be helpful where logging isn't performed.  There are currently a couple similar scenarios where raw errors are returned.

```suggestion
		return nil, fmt.Errorf("could not create Mongo transport http client: %w", err)
```

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go (hunk 3 above)
Comment (zenmoto) [reply receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:81]: I was just trying to toss in a metric here too, but I was having a difficult time finding an example of another receiver that does this. I gave myself a TODO for the moment while I look for good examples to emulate. I'll add a metric on the next PR with the client as I get it figured out. It should probably be a sum error count with the failed call as an attribute?

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go (hunk 5 above)
Comment (zenmoto) [reply receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go]: Thanks for this. I think I have all the current errors wrapped in the right places now.

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go (hunk 3 above)
Comment (jpkrohling) [reply receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:81]: Just keep this as a todo somewhere, as I'm not sure we have sorted out observability for the collector yet (or have we, @bogdandrutu?)

File: receiver/mongodbatlasreceiver/factory.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+func createMetricsReceiver(
+	ctx context.Context,
+	params component.ReceiverCreateSettings,
+	rConf config.Receiver,
+	consumer consumer.Metrics,
+) (component.MetricsReceiver, error) {
+	cfg := rConf.(*Config)
+	ms, err := newMongoDBAtlasReceiver(ctx, params.Logger, cfg, consumer)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create a MongoDB Atlas Receiver instance: %w", err)
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/factory.go:49]:
```suggestion
		return nil, errors.Wrap(err, "unable to create a MongoDB Atlas Receiver instance")
```

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+func NewMongoDBAtlasClient(
+	publicKey string,
+	privateKey string,
+	log *zap.Logger,
+) (*MongoDBAtlasClient, error) {
+	t := digest.NewTransport(publicKey, privateKey)
+	tc, err := t.Client()
+	if err != nil {
+		return nil, fmt.Errorf("Could not create MongoDB Atlas transport HTTP client: %w", err)
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:41]:
```suggestion
		return nil, errors.Wrap(err, "Could not create MongoDB Atlas transport HTTP client")
```

File: receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	pageNum int,
+) ([]*mongodbatlas.Organization, bool, error) {
+	orgs, response, err := s.client.Organizations.List(ctx, &mongodbatlas.OrganizationsListOptions{
+		ListOptions: mongodbatlas.ListOptions{
+			PageNum: pageNum,
+		},
+	})
+	err = checkMongoDBClientErr(err, response)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error in retrieving organizations: %w", err)
```
Comment (jpkrohling) [inline receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:102]:
```suggestion
		return nil, false, errors.Wrap(err, "error in retrieving organizations")
```

**Review by jpkrohling — APPROVED** [review]
LGTM, just a couple of minor things that can be improved in a follow-up PR.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**